okhttp-tests/src/test/java/okhttp3/EventListenerTest.java

```diff
@@ -523,6 +525,84 @@
     assertFalse(recordedEvents.contains(SecureConnectEnd.class));
   }
 
+  @Test public void successfulConnectionFound() throws IOException {
+    server.enqueue(new MockResponse());
+
+    Call call = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .build());
+    Response response = call.execute();
+    assertEquals(200, response.code());
+    response.body().close();
+
+    ConnectionFound connectionFound = listener.removeUpToEvent(ConnectionFound.class);
+    assertSame(call, connectionFound.call);
+    assertNotNull(connectionFound.connection);
+  }
+
+  @Test public void noConnectionFoundOnFollowUp() throws IOException {
+    server.enqueue(new MockResponse()
+        .setResponseCode(301)
+        .addHeader("Location", "/foo"));
+    server.enqueue(new MockResponse()
+        .setBody("ABC"));
+
+    Call call = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .build());
+    Response response = call.execute();
+    assertEquals("ABC", response.body().string());
+
+    listener.removeUpToEvent(ConnectionFound.class);
+
+    List<Class<?>> remainingEvents = listener.recordedEventTypes();
+    assertFalse(remainingEvents.contains(ConnectionFound.class));
+  }
+
+  @Test public void pooledConnectionFound() throws IOException {
+    server.enqueue(new MockResponse());
+    server.enqueue(new MockResponse());
+
+    // Seed the pool.
+    Call call1 = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .build());
+    Response response1 = call1.execute();
+    assertEquals(200, response1.code());
+    response1.body().close();
+
+    ConnectionFound connectionFound1 = listener.removeUpToEvent(ConnectionFound.class);
+    listener.clearAllEvents();
+
+    Call call2 = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .build());
+    Response response2 = call2.execute();
+    assertEquals(200, response2.code());
+    response2.body().close();
+
+    ConnectionFound connectionFound2 = listener.removeUpToEvent(ConnectionFound.class);
+    assertSame(connectionFound1.connection, connectionFound2.connection);
+  }
+
+  @Test public void multipleConnectionsFoundForSingleCall() throws IOException {
+    server.enqueue(new MockResponse()
+        .setResponseCode(301)
+        .addHeader("Location", "/foo")
+        .addHeader("Connection", "Close"));
+    server.enqueue(new MockResponse()
+        .setBody("ABC"));
+
+    Call call = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .build());
+    Response response = call.execute();
+    assertEquals("ABC", response.body().string());
+
+    listener.removeUpToEvent(ConnectionFound.class);
+    listener.removeUpToEvent(ConnectionFound.class);
```

Comment:
@dave-r12 Just following along from the above 3 unit tests, I see that you haven't ended the test with any assertions on the `Call`.  Was that intentional?

(I follow this codebase for learning purposes, so just looking to understand a little better.)